packages/plugins/users-permissions/server/services/providers-registry.js

```diff
@@ -2,6 +2,50 @@
 
 const { strict: assert } = require('assert');
 const jwt = require('jsonwebtoken');
+const jwkToPem = require('jwk-to-pem');
+
+const getCognitoPayload = async (idToken, purest) => {
+  const {
+    header: { kid },
+    payload,
+  } = jwt.decode(idToken, { complete: true });
+
+  if (!payload || !kid) {
+    throw new Error('The provided token is not valid');
+  }
+
+  const { iss } = payload;
+
+  const config = {
+    cognito: {
+      discovery: {
+        origin: `${iss}/.well-known/jwks.json`,
+        path: '',
+      },
+    },
+  };
+  try {
+    const cognito = purest({ provider: 'cognito', config });
+    // get the JSON Web Key (JWK) for the user pool
+    const { body: jwk } = await cognito('discovery').request();
+    // Get the key with the same Key ID as the provided token
+    const key = jwk.keys.find(({ kid: jwkKid }) => jwkKid === kid);
```

Comment:
You might want to check if the data returned by `cognito('discovery').request()` is the correct one, before accessing `keys` attribute (and `find`method). 👍

Reply:
The only way that jwk would be incorrect is if there is an error getting the key, and it would mean that there was an error verifying the provided token. So the try/catch will handle the errors inside already. What do you think?